pkg/controller/direct/discoveryengine/datastore_controller.go

```diff
@@ -0,0 +1,280 @@
+	var opts []option.ClientOption
+
+	config := m.config
+
+	// Workaround for an unusual behaviour (bug?):
+	//  the service requires that a quota project be set
+	if !config.UserProjectOverride || config.BillingProject == "" {
+		config.UserProjectOverride = true
+		config.BillingProject = projectID
+	}
+
+	opts, err := config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := gcp.NewDataStoreRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building discoveryengine datastore client: %w", err)
+	}
+
+	return gcpClient, err
+}
+
+func (m *dataStoreModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.DiscoveryEngineDataStore{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewDiscoveryEngineDataStoreIDFromObject(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := DiscoveryEngineDataStoreSpec_ToProto(mapCtx, &obj.Spec)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	return &dataStoreAdapter{
+		model:   m,
+		id:      id,
+		desired: desired,
+	}, nil
+}
+
+func (m *dataStoreModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	log := klog.FromContext(ctx)
+	if strings.HasPrefix(url, "//discoveryengine.googleapis.com/") {
+		id, err := krm.ParseDiscoveryEngineDataStoreExternal(url)
+		if err != nil {
+			log.V(2).Error(err, "url did not match DiscoveryEngineDataStore format", "url", url)
+		} else {
+			return &dataStoreAdapter{
+				model: m,
+				id:    id,
+			}, nil
+		}
+	}
+	return nil, nil
+}
+
+type dataStoreAdapter struct {
+	model   *dataStoreModel
+	id      *krm.DiscoveryEngineDataStoreID
+	desired *pb.DataStore
+	actual  *pb.DataStore
+}
+
+var _ directbase.Adapter = &dataStoreAdapter{}
+
+func (a *dataStoreAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting discoveryengine datastore", "name", a.id)
+
+	gcpClient, err := a.model.client(ctx, a.id.ProjectID)
+	if err != nil {
+		return false, err
+	}
+	req := &pb.GetDataStoreRequest{Name: a.id.String()}
+	actual, err := gcpClient.GetDataStore(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting discoveryengine datastore %q from gcp: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+func (a *dataStoreAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating discoveryengine datastore", "name", a.id)
+
+	desired := direct.ProtoClone(a.desired)
+	desired.Name = a.id.String()
+
+	gcpClient, err := a.model.client(ctx, a.id.ProjectID)
+	if err != nil {
+		return err
+	}
```

Comment:
I'm wondering why we initialize the gcpClient in each CRUD call rather than put it under `AdapterForObject`

Reply:
It's because the need for the projectID because the API doesn't auto-infer the project.  But you're right, we have the projectId in both AdapterForObject and AdapterForURL so ... moved it.